spring-batch-infrastructure/src/main/java/org/springframework/batch/item/data/builder/RepositoryItemWriterBuilder.java

```diff
@@ -0,0 +1,162 @@
+		this.repository = repository;
+
+		return this;
+	}
+
+
+	/**
+	 * Specifies a repository and the type-safe method to call for the writer.
+	 * This method can be used in place of {@link #methodName(String)}
+	 * and {@link #repository(CrudRepository)}. The
+	 * repository that is used by the repositoryMethodReference must be non-final.
+	 *
+	 * @param repositoryMethodReference of the used to get a repository and type-safe
+	 * method for use by the writer.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemWriter#setMethodName(String)
+	 * @see RepositoryItemWriter#setRepository(CrudRepository)
+	 *
+	 */
+	public RepositoryItemWriterBuilder<T> repository(RepositoryMethodReference repositoryMethodReference) {
+		this.repositoryMethodReference = repositoryMethodReference;
+
+		return this;
+	}
+
+	/**
+	 * Builds the {@link RepositoryItemWriter}.
+	 *
+	 * @return a {@link RepositoryItemWriter}
+	 */
+	public RepositoryItemWriter<T> build() {
+		if (this.repositoryMethodReference != null) {
+			this.methodName = this.repositoryMethodReference.getMethodName();
+			this.repository = this.repositoryMethodReference.getRepository();
+		}
+
+		Assert.hasText(this.methodName, "methodName is required.");
+		Assert.notNull(this.repository, "repository is required.");
+
+		RepositoryItemWriter<T> writer = new RepositoryItemWriter<>();
+		writer.setMethodName(this.methodName);
+		writer.setRepository(this.repository);
+		return writer;
+	}
+
+	/**
+	 * Establishes a proxy that will capture a the Repository and the associated
+	 * methodName that will be used by the writer.
+	 * @param <T> The type of repository that will be used by the writer.
+	 */
+	public static class RepositoryMethodReference<T> {
+		private RepositoryMethodIterceptor repositoryInvocationHandler;
+
+		private CrudRepository<?, ?> repository;
+
+		public RepositoryMethodReference(CrudRepository<?, ?> repository) {
+			this.repository = repository;
+			this.repositoryInvocationHandler = new RepositoryMethodIterceptor();
+		}
+
+		/**
+		 * The proxy returned prevents actual method execution and is only used to gather,
+		 * information about the method.
+		 * @return T is a proxy of the object passed in in the constructor
+		 */
+		public T methodIs() {
+			Enhancer enhancer = new Enhancer();
+			enhancer.setSuperclass(this.repository.getClass());
+			enhancer.setCallback(this.repositoryInvocationHandler);
+			return (T) enhancer.create();
+		}
+
+		public CrudRepository<?, ?> getRepository() {
+			return this.repository;
+		}
+
+		public String getMethodName() {
+			return this.repositoryInvocationHandler.getMethodName();
+		}
+
+	}
+
+	private static class RepositoryMethodIterceptor implements MethodInterceptor {
+		private String methodName;
+
+		@Override
+		public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy) throws Throwable {
+			this.methodName = method.getName();
+			return null;
```

Comment:
We need to add a test where the repository returns a primitive (`int` for example).  I'm wondering if this will break.  If it does, we need to document that limitation.